### PR TITLE
Entities

### PR DESCRIPTION
**Pull request by Eshchock1**
plz do not merge

### CODE REVIEW

File: src/main/java/entities/User.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package entities;
+
+public abstract class User {
+    private String username;
+    private String password;
+
```
Comment (Eshchock1) [inline src/main/java/entities/User.java:28]: missing login and logout functions

File: src/main/java/entities/Maze.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    private int numCol;
+    private char[][] state;
+    final static Map<String, Character> ENCODING = new HashMap<String, Character>() {{
+        put("empty", '.');
+        put("wall", '#');
+    }};
+
+//    public static void main(String[] args) {
+//        System.out.println(ENCODING);
+//    }
```
Comment (jchang0112) [inline src/main/java/entities/Maze.java:16]: delete for final merge

File: src/main/java/entities/PublishedMaze.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package entities;
+
+import java.time.LocalDate;
+
+public class PublishedMaze extends Maze{
+    String author;
+    String name;
+    Boolean published;
```
Comment (jchang0112) [inline src/main/java/entities/PublishedMaze.java:8]: delete?

Comment (Eshchock1) [inline src/main/java/entities/PublishedMaze.java:8]: not needed anymore, can just use instanceOf

File: src/main/java/entities/PublishedMaze.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package entities;
+
+import java.time.LocalDate;
+
+public class PublishedMaze extends Maze{
+    String author;
+    String name;
```
Comment (Eshchock1) [inline src/main/java/entities/PublishedMaze.java:7]: make this id instead of name

File: src/main/java/entities/PublishedMaze.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+    Boolean published;
+    LocalDate publishDate;
```
Comment (Eshchock1) [inline src/main/java/entities/PublishedMaze.java:9]: make this java 'Date' class instead so it includes milliseconds as well 
https://www.tutorialspoint.com/java/java_date_time.htm

File: src/main/java/entities/User.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package entities;
+
+public abstract class User {
+
+    private String user;
```
Comment (Eshchock1) [inline src/main/java/entities/User.java]: username

File: src/main/java/entities/User.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    private String password;
+
+    public User(String username, String password){
+        this.user = username;
```
Comment (Eshchock1) [inline src/main/java/entities/User.java]: username again here

File: src/main/java/entities/User.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+        this.password = password;
+    }
+
+    public abstract void userLogin();
+
+    public abstract void userLogout();
+
+
+    public String getUser(){
+        return this.user;
```
Comment (Eshchock1) [inline src/main/java/entities/User.java]: same here

File: src/main/java/entities/PublishedMaze.java (hunk 4 above)
Comment (edweenie123) [reply src/main/java/entities/PublishedMaze.java:7]: right after published maze is created, create an unique id for the published maze somehow.

one idea on how to generate the id is to concatenate the hash codes of the name of the author and the the local data object

File: src/main/java/entities/User.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+        this.user = username;
+        this.password = password;
+    }
+
+    public abstract void userLogin();
+
+    public abstract void userLogout();
+
+
+    public String getUser(){
```
Comment (Eshchock1) [inline src/main/java/entities/User.java]: getUsername

File: src/main/java/entities/PublishedMaze.java (hunk 4 above)
Comment (edweenie123) [reply src/main/java/entities/PublishedMaze.java:7]: then within the data base, we can store the id's of the mazes created by each user

File: src/main/java/entities/Player.java [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package entities;
+
+public class Player extends User{
+
```
Comment (Eshchock1) [inline src/main/java/entities/Player.java:4]: implement login and logout abstract function

Comment (Eshchock1) [inline src/main/java/entities/Player.java:4]: add the following attributes:
- prevPlayedMazes: a string array of ids of played mazes
- addPlayedMaze: a function to add a new maze id to the prevPlayedMaze array

File: src/main/java/entities/Designer.java [hunk 11]
```diff
@@ -0,0 +1,5 @@
+package entities;
+
+public class Designer extends User{
+
```
Comment (Eshchock1) [inline src/main/java/entities/Designer.java:4]: implement login and logout abstract functions

Comment (Eshchock1) [inline src/main/java/entities/Designer.java:4]: final string userType = 'designer'

File: src/main/java/entities/Player.java (hunk 10 above)
Comment (Eshchock1) [inline src/main/java/entities/Player.java:4]: final string userType = 'player'

File: src/main/java/entities/User.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
+package entities;
+
+public abstract class User {
+
```
Comment (Eshchock1) [inline src/main/java/entities/User.java:4]: add string userType attribute here

File: src/main/java/entities/Maze.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package entities;
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class Maze {
+    private int numRows;
+    private int numCol;
```
Comment (Eshchock1) [inline src/main/java/entities/Maze.java:7]: rename to numCols

File: src/main/java/entities/DesignableMaze.java [hunk 14]
```diff
@@ -0,0 +1,5 @@
+package entities;
+
+public class DesignableMaze extends Maze{
+
```
Comment (Eshchock1) [inline src/main/java/entities/DesignableMaze.java:4]: add attributes: 
- addBlock(location): which adds a wall to the maze's state at the specified location 
- removeBlock(location): which removes a wall to the maze's state at the specified location

where location is a 2 element array of integers e.g.: [row, column]

File: src/main/java/entities/GameState.java [hunk 15]
```diff
@@ -0,0 +1,5 @@
+package entities;
+
+public class GameState extends PublishedMaze{
+
```
Comment (Eshchock1) [inline src/main/java/entities/GameState.java:4]: add attributes;
- position: 2 element int array to represent x and y coordinate
- getPosition: function that returns position
- movePosition: function that will update position

File: src/main/java/entities/Maze.java [hunk 16]
```diff
@@ -0,0 +1,18 @@
+package entities;
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class Maze {
+    private int numRows;
```
Comment (Eshchock1) [inline src/main/java/entities/Maze.java:6]: need a getState function to return the state

File: src/main/java/entities/PublishedMaze.java (hunk 4 above)
Comment (Eshchock1) [reply src/main/java/entities/PublishedMaze.java:7]: Have name and id

**Pull request closed without merging**